MSC3061: Sharing room keys for past messages

uhoreg opened the pull request:
[Rendered](https://github.com/uhoreg/matrix-doc/blob/shareable_keys_msc/proposals/3061-shareable-room-keys.md)

Implementations:
- [JS SDK](https://github.com/matrix-org/matrix-js-sdk/pull/1640)
- [Polyjuice Client Test](https://gitlab.com/polyjuice/polyjuice_client_test/-/blob/develop/lib/polyjuice/client_test/test/historical_key_sharing.ex)
- [Element iOS](https://github.com/vector-im/element-ios/issues/4947)
- [Element Android](https://github.com/vector-im/element-android/issues/4153)

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/3061#issuecomment-1594770839)

ara4n commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,78 @@
+  that have this flag with the invited user.
+- when the user receives a key share request, they may share the key with the
+  requester if the user is a current member of the room.  The key may be shared
+  from the first available ratchet index, not just the requested index.
```
we should mandate that megolm ratchets are rotated when history viz settings change, i think?

uhoreg replied:
This done on lines 38-41

poljar commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,78 @@
+set to `invited` since other members generally do not send new members the keys
+to decrypt messages sent before they were invited or joined a room.
+
+We define a "shared-history" flag that identifies keys for messages that were
```
Could we add a rationale for why we're introducing a new flag vs using the history visibility setting as is?

uhoreg replied:
This is done in the "Alternatives" section

poljar replied:
Sorry I wasn't clear there. I meant to ask why aren't we storing the history visibility setting with the key.

uhoreg replied:
Ah, ok.  I can add that.

dkasak replied:
In case it's still not clear, I think the question here is why we are coercing to a boolean instead of encoding it as one of `world_readable`, `shared`, `invited`. I don't see a downside with that approach and the upside is that we're not losing any information if the difference should ever become important (e.g. if we ever add a fancy new history visibility setting in the future).

uhoreg replied:
I've added a comment in the "Alternatives" section, so I'm closing this thread.  Any discussion about which option is better should be done on a thread in that section.

uhoreg commented:
Implemented by https://github.com/matrix-org/matrix-js-sdk/pull/1640

cvwright commented:
> Implemented by [matrix-org/matrix-js-sdk#1640](https://github.com/matrix-org/matrix-js-sdk/pull/1640)

Is anyone working on this for matrix-ios-sdk?

ara4n commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,78 @@
+  share the megolm session from the first available index, rather than from the
+  current ratchet index.
+
+## Potential issues
```
This doesn't work very well with restricted rooms (MSC #3083), which are likely to be E2EE but by definition users can join them without being invited.  So we'd rely purely on keyshare reqs to backfill history, which is unfortunate.

uhoreg replied:
added a paragraph about this

deepbluev7 commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,78 @@
+## Security considerations
+
+Clients should still ensure that keys are only shared with authorized users and
+devices.
```
How does this work? How does the client know, that a user should be in the room and wasn't maliciously added by the homeserver? Even if you set a room to shared history, you probably only want members to read the history, that you accepted as having been allowed into the room. Considering the end aspect in e2ee, we want the client to approve of a member having joined the room, before sharing history. We don't want to automatically share keys with anyone, who joins the room, because the server could for a short time "fake" a user in the room and as such get access to all the history and then make the user disappear again. Then encrypting the room would be useless. Should we require prompting the user once before sharing past history with a new user in the room?

uhoreg replied:
This is sort of a generic "make sure that you don't accidentally over-share when you implement this".  In this case, clients only share the keys with users that the client invites, so it's decided that the user should be authorized.  For determining which devices are authorized, in general, the client should probably just do what it normally does for determining which devices to send keys to.

deepbluev7 replied:
I think this is a really scary trap, because a client dev might forget, that the servers device list and member list isn't really that trustworthy. We somewhat get around that by warning for unverified devices, when sending a message, but that will never work for sharing past message keys.

uhoreg replied:
I've added some more words here.  Hopefully it's clearer now.

richvdh commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,78 @@
+client may use any mechanism it wants to keep track of this flag.  When a room
+key is marked as such:
+
+- `m.room_key` and `m.forwarded_room_key` messages used to share this key have
+  a `shared_history` property set to `true`,
+- the `session_data` field in key backups of this key has a `shared_history`
+  property set to `true` in the decrypted JSON structure, and
+- the `SessionData` type used in key exports has a `shared_history` property
+  that is set to `true` for this key.
```
it would be nice to give example events etc with the new property.

richvdh commented:
Given that this is now implemented in a bunch of clients, can we land it?

uhoreg commented:
> Given that this is now implemented in a bunch of clients, can we land it?

I think that I've addressed all the comments, so it should be pretty much OK to land now.  But given that some of the comments were non-trivial, I don't think it's fair to try to push for FCP when many people are on holiday.  So I'll probably give it a few weeks into the New Year, and if it still looks fine, I'll make the FCP call then.

ara4n commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,152 @@
+The suggested behaviour in this MSC is to only share additional keys when
+inviting another user.  This does not allow users who join the room but were
+not invited (for example, if membership is restricted to another space, or if
+the room is publicly joinable) to receive the keys.  Also, if the inviter does
```
Perhaps worth noting that a quick workaround for this would be knocking: if rather than making rooms restricted-join, instead we had a restricted-knock, then users in the permitted group would be allowed to knock... and whoever answers the knock would then have the responsibility to share all the keys with them.

clokep replied:
MSC3787 (in room version 10) added a `knock_restricted` join rule which lets users knock or join via the join rule restrictions. I think you're suggesting allowing them to *knock* if they meet the restrictions though?

kristovatlas commented:
I see client implementations but not server ones. Does this require making server changes as well or can it be implemented purely client-side?

uhoreg commented:
It's entirely client-side.

bradtgmurray commented:
Shall we bump this up and land this now that a few weeks since new years has passed? :)

uhoreg commented:
@mscbot fcp merge

richvdh commented on proposals/3061-shareable-room-keys.md:
```diff
@@ -0,0 +1,152 @@
+share keys with users when the client invites them, as the client is then
+certain that the user is allowed to be in the room.  Another way is to have a
+mechanism of verifying membership, such as the method proposed in
+(MSC3917)[https://github.com/matrix-org/matrix-spec-proposals/pull/3917].
```
```suggestion
[MSC3917](https://github.com/matrix-org/matrix-spec-proposals/pull/3917).
```

turt2live commented:
The [FCP](https://github.com/matrix-org/matrix-spec-proposals/pull/3061#issuecomment-1594770839) for this MSC recently has had enough checkboxes, but the bot didn't start FCP due to a bug. 

🔔 This MSC is now entering FCP, as of the timestamp of this comment. 🔔

turt2live commented:
This MSC has completed FCP per the [above comment](https://github.com/matrix-org/matrix-spec-proposals/pull/3061#issuecomment-1671589660).

uhoreg commented:
Spec PR: https://github.com/matrix-org/matrix-spec/pull/1655

HarHarLinks commented:
Removed from matrix-js-sdk in https://github.com/matrix-org/matrix-js-sdk/releases/tag/v34.8.0

richvdh commented:
For links: this proposal is being replaced by [MSC4268](https://github.com/matrix-org/matrix-doc/pull/4268).
